### ISSUE TITLE
feat(demo): built-in markdown plugin provides support for showcasing demo capabilities.

### ISSUE THREAD

**Issue by hairyf** [vuejs/vitepress#2432]
### Is your feature request related to a problem? Please describe.

vitepress is currently not capable enough for developing component libraries. While it is fast, it lacks something. We may be able to incorporate support for component capabilities similar to `dumi` to help developers quickly develop Vue component libraries docs.

```html
<demo src="../demo.vue" title="Demo block" desc="use demo" />
```

Alternatively, use the `desc` defined in the `container` to write Markdown syntax.

```md
::: demo src="../demo.vue" title="Demo block"

This is a `description` that can be written using Markdown.

:::
```

### Describe the solution you'd like

To address this, I have created the [markdown-it-vitepress-demo](https://github.com/hairyf/markdown-it-vitepress-demo) plugin. I'm not sure if it can be helpful.

### Additional context

- https://github.com/vuejs/vitepress/issues/987
- https://github.com/vuejs/vitepress/issues/1349

### Validations

- [X] Follow our [Code of Conduct](https://vuejs.org/about/coc.html)
- [X] Read the [docs](https://vitepress.dev).
- [X] Read the [Contributing Guidelines](https://github.com/vuejs/vitepress/blob/main/.github/contributing.md).
- [X] Check that there isn't already an issue that asks for the same feature to avoid creating a duplicate.

**Comment by brc-dd** [comment]
Hey, is there a deployed site we can see that plugin in action?

**Comment by hairyf** [comment]
of course, I have built a component library using that plugin at my company. It's not very convenient to showcase, but I can provide a simple demo implementation:

https://markdown-it-vitepress-demo.vercel.app/api-examples.html

The implementation can be found within the markdown-it-vitepress-demo sandbox, and you can view the source code here:

https://github.com/hairyf/markdown-it-vitepress-demo/tree/main/docs

Additionally, I have created a similar demo component based on [markdown-it-vitepress-demo](https://github.com/hairyf/markdown-it-vitepress-demo), resembling naive-ui. However, I'm not ready to release it yet.
![2G4SBOXG_J5V41Z~JCXY{YG](https://github.com/vuejs/vitepress/assets/49724027/cdf0ebdb-0ab2-419d-a72d-421a87e04208)

I believe VitePress should have a uniquely styled built-in default `demo-container` component, which can also be overridden by the user.

**Comment by dwgray** [comment]
This seems like an essential feature for building documentation sites for vue.js libraries. Is there an alternative solution that I'm missing, or are folks just working around this for now? It seems like @hairyf 's plugin is a good start, but it hasn't been touched for a year and doesn't have the markers of an extensively used project.  I'd be happy to help with a solution for this if someone can point me in the right direction (or would be even happier if there is an existing alternative that I'm missing).

My definition of 'this': I'm contributing to a vue.js 3x library - [bootstrap-vue-next](https://github.com/bootstrap-vue-next/bootstrap-vue-next) that is a rewrite of a vue.js 2x library [bootstrap-vue](https://bootstrap-vue.org/). We're generally trying to use the newest toolset, so long before I had any involvement in this project, the owners opted for vitepress over vuepress. In the old project, when documenting usage of one of the components, all that was required was a straightforward implementations of the sample code set off in an ` ```html ` block (obviously, there is a lot going on under the hood, but there isn't anything else required in the md file - most specifically there isn't any duplication of the sample code or scaffolding).

[Example on GitHub](https://github.com/bootstrap-vue/bootstrap-vue/blob/dev/src/components/form-checkbox/README.md?plain=1#L8-L9)

Snapshot:
![image](https://github.com/vuejs/vitepress/assets/5083020/11cec023-3f49-4969-8e5c-6ce594591b32)

What we're doing in vitepress is quite a bit uglier - not only is the sample manually duplicated, but there is a significant amount of scaffolding code that's repeated for each sample + the `<script>` part of the running sample has to be included in the `script` for the main file, which means variables have to be unique for each sample within a file.

[Example on Github](https://github.com/bootstrap-vue-next/bootstrap-vue-next/blob/main/apps/docs/src/docs/components/form-checkbox.md?plain=1#L17)

Snapshot:
![image](https://github.com/vuejs/vitepress/assets/5083020/474d3b01-c5d6-4606-bd7e-f66c4853541d)

and the following is included in the main ```script``` block
```
import { BFormCheckbox} from 'bootstrap-vue-next'
const status = ref(false);
```

Thanks in advance for any direction you can give me on this.  Again, I'd be happy to help with a solution for this if someone can point me in the right direction.

**Comment by hairyf** [comment]
@dwgray If you don't mind, you can use markdown-it-vitepress-demo to create a demo, which is still valid in the latest version of vite

**Comment by dwgray** [comment]
Thanks, @hairyf. I'll take a closer look now that I know that you're actively using it in your vue3-pixi.
